src/main/java/com/es/phoneshop/web/ProductDetailsPageServlet.java

```diff
@@ -1,33 +1,71 @@
 package com.es.phoneshop.web;
 
 import com.es.phoneshop.model.product.ArrayListProductDao;
+import com.es.phoneshop.model.product.Cart.Cart;
+import com.es.phoneshop.model.product.Cart.CartService;
+import com.es.phoneshop.model.product.Cart.HttpSessionCartService;
+import com.es.phoneshop.model.product.Product;
+import com.es.phoneshop.model.product.exceptions.OutOfStockException;
 import com.es.phoneshop.model.product.exceptions.ProductNotFoundException;
+import com.es.phoneshop.model.product.recentlyViewed.RecentlyViewedService;
+import com.es.phoneshop.model.product.recentlyViewed.HttpSessionRecentlyViewedService;
 
 
 import javax.servlet.ServletException;
 import javax.servlet.http.HttpServlet;
 import javax.servlet.http.HttpServletRequest;
 import javax.servlet.http.HttpServletResponse;
 import java.io.IOException;
+import java.text.NumberFormat;
+import java.text.ParseException;
+import java.util.List;
 
 public class ProductDetailsPageServlet extends HttpServlet {
     private ArrayListProductDao arrayListProductDao;
+    private CartService cartService;
+    private RecentlyViewedService recentlyViewedService;
 
 
     @Override
     public void init() {
         this.arrayListProductDao = ArrayListProductDao.getInstance();
+        this.cartService = HttpSessionCartService.getInstance();
+        this.recentlyViewedService = HttpSessionRecentlyViewedService.getInstance();
     }
 
     @Override
     protected void doGet(HttpServletRequest request, HttpServletResponse response) throws ServletException, IOException {
+        List<Product> recentlyViewedProducts = recentlyViewedService.getRecentlyViewed(request);
         try {
-            Long productId = Long.valueOf(request.getPathInfo().substring(1));
-            request.setAttribute("product", arrayListProductDao.getProduct(productId));
-            request.getRequestDispatcher("/WEB-INF/pages/productDetails.jsp").forward(request, response);
+            Cart cart = cartService.getCart(request);
+            request.setAttribute("cart", cart);
+            recentlyViewedService.addRecentlyViewedProduct(recentlyViewedProducts, getProductFromPath(request).getId());
```

Comment:
Don't make an assumption that the list returned before try statement will contain the product added in this line (even though it does). Add, then obtain